engine/storage/snapshot/src/org/apache/cloudstack/storage/snapshot/XenserverSnapshotStrategy.java

```diff
@@ -260,11 +260,15 @@ public boolean deleteSnapshot(Long snapshotId) {
             boolean result = deleteSnapshotChain(snapshotOnImage);
             obj.processEvent(Snapshot.Event.OperationSucceeded);
             if (result) {
-                //snapshot is deleted on backup storage, need to delete it on primary storage
-                SnapshotDataStoreVO snapshotOnPrimary = snapshotStoreDao.findBySnapshot(snapshotId, DataStoreRole.Primary);
-                if (snapshotOnPrimary != null) {
-                    snapshotOnPrimary.setState(State.Destroyed);
-                    snapshotStoreDao.update(snapshotOnPrimary.getId(), snapshotOnPrimary);
+                SnapshotInfo snapshotOnPrimary = snapshotDataFactory.getSnapshot(snapshotId, DataStoreRole.Primary);
```

Comment:
Master branch has slightly different code which deals with cleanup for RBD type storage pool. So if this is fwd merged there will be conflicts.